lib/segment/src/index/field_index/full_text_index/inverted_index.rs

```diff
@@ -1,40 +1,66 @@
-use std::collections::{BTreeSet, HashMap};
+use std::collections::HashMap;
 
+use ahash::AHashSet;
 use common::counter::hardware_counter::HardwareCounterCell;
 use common::types::PointOffsetType;
-use serde::{Deserialize, Serialize};
+use itertools::Itertools;
 
 use crate::common::operation_error::OperationResult;
 use crate::index::field_index::{CardinalityEstimation, PayloadBlockCondition, PrimaryCondition};
 use crate::types::{FieldCondition, Match, PayloadKeyType};
 
 pub type TokenId = u32;
 
-#[derive(Default, Serialize, Deserialize, Debug, Clone)]
-pub struct Document {
-    tokens: Vec<TokenId>,
-}
+/// Contains the set of tokens that are in a document.
+///
+/// Internally, it keeps them unique and sorted, so that we can binary-search over them
+#[derive(Default, Debug, Clone)]
+pub struct TokenSet(Vec<TokenId>);
 
-impl Document {
-    pub fn new(mut tokens: Vec<TokenId>) -> Self {
-        tokens.sort_unstable();
-        Self { tokens }
+impl TokenSet {
+    pub fn new(tokens: AHashSet<TokenId>) -> Self {
+        let sorted_unique = tokens.into_iter().sorted_unstable().collect();
+
+        Self(sorted_unique)
     }
 
     pub fn len(&self) -> usize {
-        self.tokens.len()
+        self.0.len()
     }
 
     pub fn is_empty(&self) -> bool {
-        self.tokens.is_empty()
+        self.0.is_empty()
     }
 
     pub fn tokens(&self) -> &[TokenId] {
-        &self.tokens
+        &self.0
+    }
+
+    pub fn contains(&self, token: &TokenId) -> bool {
+        self.0.binary_search(token).is_ok()
     }
+}
+
+impl FromIterator<TokenId> for TokenSet {
+    fn from_iter<T: IntoIterator<Item = TokenId>>(iter: T) -> Self {
+        let tokens = iter
+            .into_iter()
+            .sorted_unstable()
+            .dedup()
+            .collect::<Vec<_>>();
 
-    pub fn check(&self, token: &TokenId) -> bool {
-        self.tokens.binary_search(token).is_ok()
+        Self(tokens)
+    }
+}
+
+/// Contains the token ids that make up a document, in the same order that appear in the document.
```

Comment:
Also note in the comment that this can contain duplicates, in contrast to the `TokenSet`.